### ISSUE TITLE
redefined-outer-name using with statments

### ISSUE THREAD

**Issue by FantasticMrFux** [PyCQA/pylint#5375]
### Bug description

Consider this python script.

``` python
#!/usr/bin/python3
"""Docstring."""

with open("./parameters.txt" , "r", encoding="UTF-8") as file_handle:
    paramters = file_handle.split()


def main():
    """Docstring."""
    with open("./output.txt" , "w", encoding="UTF-8") as file_handle:
        file_handle.write("Output")


if __name__ == "__main__":
    main()
```

I do not understand why I should not be able to reuse the variable name when it is only in the "scope" of the with-statement. The same is true for variable name defined inside of a comprehension. 

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:10:57: W0621: Redefining name 'file_handle' from outer scope (line 4) (redefined-outer-name)

------------------------------------------------------------------
Your code has been rated at 8.57/10 (previous run: 8.57/10, +0.00)
```

### Expected behavior

No error.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.5
Python 3.6.8 (default, Aug 24 2021, 23:39:28) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-3)]
```

### OS / Environment

CentOS

**Comment by Pierre-Sassoulas** [comment]
The context manager is not the same as a comprehension. ``file_handle`` exists in the global scope after the ``with`` (you can try to print it it works). Consider the following file ``a.py``:
````
with open("./parameters.txt" , "r", encoding="UTF-8") as file_handle:
    paramters = file_handle
print(file_handle) # <_io.TextIOWrapper name='./parameters.txt' mode='r' encoding='UTF-8'>

t = {i:"j" for i in (1,2,3)}
# print(i) # Crash: NameError: name 'i' is not defined

def f():
    i = 0 # nothing
    file_handle = 1 #(redefined-outer-name)
````

Pylint output:
```
************* Module a
a.py:10:4: W0621: Redefining name 'file_handle' from outer scope (line 1) )
```

**Comment by FantasticMrFux** [comment]
To be honest I was not aware before I saw the error from pylint. Then I played and figured that out. Basically doing what you did. Thats why I put scope in quotes.

Yet is not clear the the as-variable should be only used inside of the with-statment? Would there be any reason for using it after the with-statment?
In contrast to a for-loop where you might want to know the last value of the iterator.

As such I would have thought this is could be called a "soft-scope" or such. And if this all is true than there should be no error for a reuse of a as-variable. And further shouldn't the code you showed not give a different error? For the reuse of the as-variable out side of the "soft-scope" of the with-statment.

If you could help me with this I would be grateful.
Thanks.

**Comment by Pierre-Sassoulas** [comment]
There's the same problem with for loop as you noted. My opinion is that this is a python design decision, and we, as a statistical analysis tool, are not in position to determine if the variable still existing out of its "intended scope" is a good thing or a bad thing, it just is the case in python. Starting to add opinionated special cases to our checkers would increase the maintenance burden by a lot. We'd have to maintain this special case (and a hundred unanticipated variation of it). And then argue with those that oppose the special case, and they will have python's design decisions behind their arguments. Doing nothing as long as python do not change is cheaper and make more sense whatever my opinion on the issue is.

**Comment by FantasticMrFux** [comment]
Okay thanks for you elaboration.
